juju/juju
# Use lease management for singular controller workers

## Description of change

The aim of this PR is to get rid of Mongo mastership as a means of running singular controller workers, and while doing so, move several more state workers into the machine manifold (txnpruner, dblogpruner).

The "singular" lease manager is updated to manage leases for either the controller or model. The machine manifolds contain a singular worker that runs for all controller agents, and attempts to claim a lease on the controller. The txnpruner, dblogpruner, and externalcontrollerupdater manifolds/workers depend on that to start up.

We introduce two new flags to the machine manifolds: is-controller-flag, and is-responsible-controller-flag. The former is used to indicate that the agent is a controller, and the latter indicates that the agent has claimed responsibility for the controller.

## QA steps

1. juju bootstrap localhost
2. juju enable-ha
(wait for all)
3. juju deploy ubuntu
4. check that only machine 0 in the controller model is running "external-controller-updater", "log-pruner", and "transaction-pruner" (use the juju-engine-report command)
5. juju ssh -m controller 0 sudo service jujud-machine-0 stop
6. wait for another agent to claim the controller lease, again checking only one agent runs the workers specified above

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1726680
